File: leetcode/MaximumSubarray.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# ===============================================================
#
# Filename:	MaximumDepthOfBinaryTree.py
#
# Author:		Oxnz
# Email:		yunxinyi@gmail.com
# Created:		[2014-12-02 17:33:41 CST]
# Last-update:	2014-12-02 17:33:41 CST
# Description: ANCHOR
#
# Version:		0.0.1
# Revision:	[None]
# Revision history:	[None]
# Date Author Remarks:	[None]
#
# License:
# Copyright (c) 2013 Oxnz
#
# Distributed under terms of the [LICENSE] license.
# [license]
#
# ===============================================================
#

class Solution:
    def maxSubArray(self, A):
        s = A[0]
        t = A[0]
        for n in A[1:]:
            s = max(s + n, n)
            t = max(t, s)
        return t

    def maxSubArray2(self, A):
        m = None
        s = 0
        t = 0
        for n in A:
            s += n
            m = max(m, s)
            if n < 0:
                t = 0
            else:
                t += n
            if s < 0:
                s = t
        return m


import unittest

class TestMaximumDepthOfBinaryTree(unittest.TestCase):
    def setUp(self):
        self._maxSubArray = Solution().maxSubArray

    def test_maxSubArray(self):
        self.assertEqual(1, self._maxSubArray([1]))
        self.assertEqual(-1, self._maxSubArray([-1]))
        self.assertEqual(6, self._maxSubArray([-2,1,-3,4,-1,2,1,-5,4]))

if __name__ == '__main__':
    unittest.main()

File: leetcode/MaximumSubarray.java
/*
 * Filename:	MaximumSubarray.java
 *
 * Author:		Oxnz
 * Email:		yunxinyi@gmail.com
 * Created:		[2014-12-02 21:20:08 CST]
 * Last-update:	2014-12-02 21:20:08 CST
 * Description: anchor
 *
 * Version:		0.0.1
 * Revision:	[NONE]
 * Revision history:	[NONE]
 * Date Author Remarks:	[NONE]
 *
 * License: 
 * Copyright (c) 2013 Oxnz
 *
 * Distributed under terms of the [LICENSE] license.
 * [license]
 *
 */

/*
this problem was discussed by Jon Bentley (Sep. 1984 Vol. 27 No. 9 Communications of the ACM P885)

the paragraph below was copied from his paper (with a little modifications)

algorithm that operates on arrays: it starts at the left end (element A[1]) and scans through to the right end (element A[n]), keeping track of the maximum sum subvector seen so far. The maximum is initially A[0]. Suppose we've solved the problem for A[1 .. i - 1]; how can we extend that to A[1 .. i]? The maximum sum in the first I elements is either the maximum sum in the first i - 1 elements (which we'll call MaxSoFar), or it is that of a subvector that ends in position i (which we'll call MaxEndingHere).

MaxEndingHere is either A[i] plus the previous MaxEndingHere, or just A[i], whichever is larger.

public static int maxSubArray(int[] A) {
    int maxSoFar=A[0], maxEndingHere=A[0];
    for (int i=1;i<A.length;++i){
        maxEndingHere= Math.max(maxEndingHere+A[i],A[i]);
        maxSoFar=Math.max(maxSoFar, maxEndingHere); 
    }
    return maxSoFar;
}

*/
public class MaximumSubarray {
    public MaximumSubarray {
    }

    public static void main(String[] args) {
        System.out.println("Hello world!");
    }
}

